### DISCUSSION TITLE
Possible to compose business & tech roles in deeper hierarchy?

### DISCUSSION THREAD

**Discussion by acaruso7** [littleK0i/SnowDDL#74] [Q&A]
I'm in the process of migrating my Snowflake IaC repo from Terraform --> SnowDDL, and I had originally planned out a pretty deep hierarchy of role inheritance. For example, suppose we had a `loader` business role which is a member of a `raw_db_admin` role. That `raw_db_admin` role might also be a member of a `raw_db_reader` and `raw_db_writer` role. And those reader and writer roles might be members of all the individual [SnowDDL-generated] schema read and write tech roles

Is this type of deep inheritance supported by the tool? It seems to me that only one level of inheritance in manually configured roles in supported (business roles can inherit from tech roles). But what if I want tech roles to inherit from other tech roles as a means of abstracting out commonly re-used sets of grants?

On one hand, I can see the benefit of avoiding deep role inheritance structures. I am coming from a Terraform deploy, and one issue was that it was challenging to actually traverse the configuration and get a good grasp on which grants a given user role actually had due to all the nested role memberships defined. On the other hand, it seems like maintainability is going to be a challenge in SnowDDL if I have to repeatedly grant business roles access to low level tech roles, when those tech roles could potentially be encapsulated into higher level abstractions for easier re-use

For reference, here's a little screenshot of an example role inheritance structure I had diagrammed out prior to starting my SnowDDL implementation. The red boxes are manually configured roles, and the green auto-generated tech roles. User roles not pictured here. This was before I really got into it and realized this might not be feasible. Am I thinking about the intended usage of this tool wrong?

<img width="956" alt="image" src="https://github.com/littleK0i/SnowDDL/assets/29766995/b448b4b8-ed31-4ba5-bd63-72384d03f073">

Looking forward to hearing others' thoughts on this!

**Comment by littleK0i** [comment]
The current 3-tier role hierarchy was designed specifically with long term maintainability in mind.

We had a few hundred "live" users and dozens of "script" users with permissions being adjusted all the time. Some permissions were generated and adjusted dynamically. We also had to pass a strict external audit. Any kind of multi-nested hierarchy in this environment would be very difficult to maintain.

---

In theory, the repetition should not be a big issue with the current setup:

1. You may specify patterns for schema names. For example, pattern `db1.*` would match all schemas in database `db1`. No need to specify each schema individually.
2. Normally you should not need a large number of "tech" roles. It is much easier to manage permissions on schema level via auto-generated "schema" roles.
3. Each user may have multiple business roles. You may combine some common permissions in designated "business" role, give it descriptive name and grant it to users in addition to more specific business roles.
4. "Schema owner" role automatically includes all permissions. No need to specify "read", "write" and "owner" simultaneously. Keeping only "owner" is sufficient.

Based on your picture, the setup could be relatively simple.

business_roles:
```yaml
loader_raw_db:
  schema_owner:
    - raw_db.*

loader_another_db:
  schema_owner:
    - another_db.*
```

users:
```yaml
airbyte_loader:
  ...
  business_roles:
    - loader_raw_db
    - loader_another_db
```

---

One final note about tech roles.

Normally you only need tech roles to grant access to very limited number of objects. For example, you have a user with very restricted access, which should only have access to a few tables.

Tech roles should generally not be used for majority of grants, since it is too verbose and becomes hard to manage when number of objects grows. Schema roles and warehouse roles should cover 95% of typical use cases.

**Reply by acaruso7** [reply]
This all makes a lot of sense. At first I wanted a deeper hierarchy, but now I realize the benefit of limiting the depth of the hierarchy. It's good to have an opinionated tool here so things don't get out of hand

I think I might still want separate business roles for something like a general purpose `loader` that has create, update, read, grant on everything in a raw db, but also a `raw_reader` that only gets read on that db. But I can just keep the role hierarchy "flat" by adding them as separate business roles, even if there's some overlap

I may have some additional questions on intended usage here as I work through my implementation, but things are clear for now. Thanks for the input!
